vignettes/reliability.Rmd

```diff
@@ -0,0 +1,347 @@
+
+$$
+\rho = \frac{(\bv a' \bv \lambda)^2 + \bv \lambda'V_{\hat a} \bv \lambda}{(\bv a' \bv \lambda)^2 + \bv a'\bv \Theta \bv a + \bv \lambda'V_{\hat a}\bv \lambda + \bv a' V_{\hat \nu} \bv a + \Tr(V_{\hat a} V_{\hat \nu}) + \Tr(\bv \Theta \bv V_{\hat a})}
+$$
+
+The sample estimates can be obtained by substituting in the sample estimates of $\bv \lambda$ and $\bv \Theta$ and using the corresponding form of $\bv a$.
+
+#### Quadratic formula for regression factor scores
+
+The reliability formula for the regression factor scores derived using a quadratic formula is
+
+$$\text{Rel} = \frac{1\pm\sqrt{1 - 4\theta_{corrected}/\psi}}{2}$$
+
+where $\theta_{corrected}$ is the corrected error term for the factor scores.
+
+## Empirical Example
+
+### Regression Factor Scores
+
+```{r}
+# Three items from the classic Holzinger and Swineford (1939) data
+cfa_fit <- cfa("
+  visual =~ x1 + x2 + x3
+",
+  data = HolzingerSwineford1939,
+  std.lv = TRUE
+)
+# Uncorrected reliability of regression factor scores
+fsr <- lavPredict(cfa_fit, fsm = TRUE, acov = TRUE)
+attr(fsr, "fsm")[[1]] %*% lavInspect(cfa_fit, what = "est")$lambda
+# Empirical reliability of regression factor scores
+var(fsr) / (var(fsr) + attr(fsr, "acov")[[1]])
+```
+
+```{r}
+# Corrected reliability of regression factor scores
+get_a <- function(lambda, theta) {
+  solve(tcrossprod(lambda) + theta, lambda)
+}
+lam <- lavInspect(cfa_fit, what = "est")$lambda
+th <- lavInspect(cfa_fit, what = "est")$theta
+jac_a <- lavaan::lav_func_jacobian_complex(
+  \(x) get_a(x[1:3], diag(x[4:6])),
+  x = c(lam, diag(th))
+)
+# Estimated sampling variance of weights
+va <- jac_a %*% vcov(cfa_fit)[1:6, 1:6] %*% t(jac_a)
+aa <- tcrossprod(get_a(lam, th)) + va
+sum(diag(tcrossprod(lam) %*% aa)) /
+  sum(diag(cfa_fit@implied$cov[[1]] %*% aa))
+# Or using `get_fs_lavaan`
+get_fs_lavaan(cfa_fit, reliability = TRUE) |>
+  attr(which = "reliability")
+```
+
+### Bartlett Factor Scores
+
+```{r}
+# Uncorrected reliability of Bartlett factor scores
+fsb <- lavPredict(cfa_fit, method = "Bartlett", fsm = TRUE, acov = TRUE)
+1 / (1 + attr(fsb, "acov")[[1]])
+# Empirical reliability of regression factor scores
+(var(fsb) - attr(fsb, "acov")[[1]]) / var(fsb)
+```
+
+```{r}
+# Corrected reliability of regression factor scores
+get_a <- function(lambda, theta) {
+  thinv_lam <- solve(theta, lambda)
+  solve(crossprod(lambda, thinv_lam), t(thinv_lam))
+}
+lam <- lavInspect(cfa_fit, what = "est")$lambda
+th <- lavInspect(cfa_fit, what = "est")$theta
+jac_a <- lavaan::lav_func_jacobian_complex(
+  \(x) get_a(x[1:3], diag(x[4:6])),
+  x = c(lam, diag(th))
+)
+# Estimated sampling variance of weights
+va <- jac_a %*% vcov(cfa_fit)[1:6, 1:6] %*% t(jac_a)
+aa <- tcrossprod(t(get_a(lam, th))) + va
+sum(diag(tcrossprod(lam) %*% aa)) /
+  sum(diag(cfa_fit@implied$cov[[1]] %*% aa))
+# Or using `get_fs_lavaan` (need to be fixed)
+R2spa::get_fs_lavaan(cfa_fit, method = "Bartlett", reliability = TRUE) |>
```

Comment:
Right now the function doesn't seem to compute the right reliability with Bartlett scores.

Reply:
Need to add a unit test as well.